Scripts/Runtime/Data/Collections/UnsafeTypedStream.cs

```diff
@@ -765,6 +825,77 @@ internal void ReadBlock(byte* dstStart, int dstStartIndex, int elementsToRead)
                 return ref UnsafeUtility.AsRef<T>(m_ReaderHead);
             }
         }
+        
+        //*************************************************************************************************************
+        // ENUMERATOR
+        //*************************************************************************************************************
+        
+        [BurstCompile]
+        public struct Enumerator : IEnumerator<T>
+        {
+            [NativeDisableUnsafePtrRestriction] private readonly BufferInfo* m_BufferInfo;
+            private int m_LaneIndex;
+            private LaneReader m_LaneReader;
+            private int m_ArrayIndex;
+            
+            public T Current
+            {
+                get => m_LaneReader.Read();
```

Comment:
Might want to cache the current value locally like they do in `List<T>`
https://referencesource.microsoft.com/#mscorlib/system/collections/generic/list.cs,cf7f4095e4de7646

This way you don't get an error if you read `Current` before calling `MoveNext` for the first time. It's at least consistent with .NET. Also seems to handle calling `Current` after you reach the end of the enumerable.

Reply:
Interesting. Everything I remember reading was that you were really not supposed to call `Current` before `MoveNext` or at the end. `Enumerators` worked on a very specific set of rules that had to be followed and that `foreach` code gen expected. 

That said, I couldn't find those readings anywhere and it does look like .NET ensures you don't error for all its enumerators so I agree, it's best to be consistent. 

I've made the change. 

There is a risk (albeit a small one) that someone comes along and does a `GetEnumerator` manually. Doesn't know the "rules" and calls `Current` first and then `MoveNext` in their loop. They will get back `default` which due to the struct nature of everything here, may look completely normal so they don't notice. They will get one extra iteration in their loop they weren't expecting but unless they are trying to add to a presized array, they won't notice that either. 

Before when it would throw an error, at least they would immediately look into it and learn how `Enumerators` work and ideally just use a `foreach` but still, they'd get the feedback right away and it would be easier to debug. 

I think the risk here is less than being consistent with .NET so it's fine, but just bringing it up for posterity.

Reply:
Agreed on the risk.
We could add a debug only log message warning the developer 🤷

Reply:
Yeah I think it's fine. I haven't seen anyone use the actual `GetEnumerator` call for a long long long time so I feel like that would set off some flags in the PR where we ask "Why no `foreach`"?